### PR TITLE
ENT-2722: Test for carriage return before newline

### PR DESCRIPTION
**Pull request by karlhto**
Just want to run this on all platforms to see if possible to recreate bug from ENT-2722. From what I can see it was fixed by @jimis in 2016.

### CODE REVIEW

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics, please!

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1345/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1345/

**Comment by karlhto** [comment]
@cf-bottom Jenkins with exotics. Want this to be tested for everything to make sure that the error is gone.

Also `csv_parser.c` is full of spaghetti

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1346/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1346/

**Comment by karlhto** [comment]
Adds test for empty fields before carriage return before newline (`\r\n`).

File: tests/unit/csv_parser_test.c [hunk 1]
```diff
@@ -215,6 +215,27 @@ static void test_new_csv_reader_zd3151_ENT3023()
...
 
+static void test_new_csv_reader_carriage_return()
+{
+    size_t len = 4;
+    const char *inputs[] =
+    {
+        "default,paths,path[awk],0,/bin/awk,source=agent\r\n",
+        "default,sys,ipv4_2[eth0],0,10.186,source=agent\r\n",
+        "default,mon,dev_temp3,0,0.00,source=agent\r\n",
+        "default,sys,ipv4_2[eth0],0,10.186,source=agent\r\n"
```
Comment (olehermanse) [inline tests/unit/csv_parser_test.c]: @karlhto It seems you are missing the line which you mentioned in PR comment. Maybe change last line to be:

```
default,sys,ipv4_2[eth0],0,10.186,\r\n
```
(comma/empty field before carriage return).

A line with only quotes before carriage return could be good as well.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1367/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1367/

**Comment by karlhto** [comment]
Started a new build:
[Build Status](https://ci.cfengine.com/job/pr-pipeline/1369/)

https://ci.cfengine.com/job/pr-pipeline/1369/

The previous one somehow never started.

**Comment by olehermanse** [comment]
@karlhto jenkins has been offline. It works now if you comment again or delete the cf-bottom comment.

**Comment by karlhto** [comment]
@olehermanse I am unable to delete the cf-bottom comment, could you do it?

**Comment by olehermanse** [comment]
@karlhto The build failed to start because you used a "non-breaking space" in the groovy expression. I think this happens when you press `alt-space` or something similar on a mac keyboard.

**Comment by karlhto** [comment]
Started new builds that previously failed:
[Build Status](https://ci.cfengine.com/job/pr-pipeline/1410/)

https://ci.cfengine.com/job/pr-pipeline/1410/

**Comment by karlhto** [comment]
@olehermanse You can remove the WIP tag now. This should be ready for merging.

**Comment by karlhto** [comment]
Solaris seems to crash? Gets a "Write failed" before exiting. It gets past `csv_parser_test` though, so `csv_parser` seems to work fine.

**Pull request merged**